api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule_service.proto

```diff
@@ -0,0 +1,97 @@
+service DatabaseObjectImportRuleService {
+  // GetDatabaseObjectImportRule is used to query a DatabaseObjectImportRule resource by its name.
+  //
+  // This will return a NotFound error if the specified DatabaseObjectImportRule does not exist.
+  rpc GetDatabaseObjectImportRule(GetDatabaseObjectImportRuleRequest) returns (DatabaseObjectImportRule);
+  // ListDatabaseObjectImportRules is used to query DatabaseObjectImportRules.
+  //
+  // Follows the pagination semantics of
+  // https://cloud.google.com/apis/design/standard_methods#list.
+  rpc ListDatabaseObjectImportRules(ListDatabaseObjectImportRulesRequest) returns (ListDatabaseObjectImportRulesResponse);
+  // CreateDatabaseObjectImportRule is used to create a DatabaseObjectImportRule.
+  //
+  // This will return an error if a DatabaseObjectImportRule by that name already exists.
+  rpc CreateDatabaseObjectImportRule(CreateDatabaseObjectImportRuleRequest) returns (DatabaseObjectImportRule);
+  // UpdateDatabaseObjectImportRule is used to modify an existing DatabaseObjectImportRule.
+  rpc UpdateDatabaseObjectImportRule(UpdateDatabaseObjectImportRuleRequest) returns (DatabaseObjectImportRule);
+  // UpsertDatabaseObjectImportRule is used to create or replace an existing DatabaseObjectImportRule.
+  //
+  // Prefer using CreateDatabaseObjectImportRule and UpdateDatabaseObjectImportRule.
+  rpc UpsertDatabaseObjectImportRule(UpsertDatabaseObjectImportRuleRequest) returns (DatabaseObjectImportRule);
+  // DeleteDatabaseObjectImportRule is used to delete a specific DatabaseObjectImportRule.
+  //
+  // This will return a NotFound error if the specified DatabaseObjectImportRule does not exist.
+  rpc DeleteDatabaseObjectImportRule(DeleteDatabaseObjectImportRuleRequest) returns (google.protobuf.Empty);
+}
+
+// The request for CreateDatabaseObjectImportRule.
+message CreateDatabaseObjectImportRuleRequest {
+  // The DatabaseObjectImportRule to create.
+  DatabaseObjectImportRule item = 1;
+}
+
+// The request for GetDatabaseObjectImportRule.
+message GetDatabaseObjectImportRuleRequest {
+  // The name of the DatabaseObjectImportRule to fetch.
+  string name = 1;
+}
+
+// The request for ListDatabaseObjectImportRules.
+message ListDatabaseObjectImportRulesRequest {
+  // The maximum number of items to return.
+  // The server may impose a different page size at its discretion.
+  int32 page_size = 1;
+  // The next_page_token value returned from a previous List request, if any.
+  string page_token = 2;
+}
+
+// The response for ListDatabaseObjectImportRules.
+message ListDatabaseObjectImportRulesResponse {
+  // The page of DatabaseObjectImportRules that matched the request.
+  repeated DatabaseObjectImportRule items = 1;
+  // Token to retrieve the next page of results, or empty if there are no
+  // more results in the list.
+  string next_page_token = 2;
+}
+
+// The request for UpdateDatabaseObjectImportRule.
+message UpdateDatabaseObjectImportRuleRequest {
```

Comment:
No update_mask on Update? I recommend you add one, it helps to prevent squashing fields we didn't want to touch in the first place.

* https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md#update
* https://cloud.google.com/apis/design/standard_methods#update

Reply:
As far as I can see `update_mask` leads to complex, brittle code that requires a ton of testing for little benefit. For example, this string comparisons is nasty:
```golang
		case path == "spec.roles":
			role.SetImpersonateConditions(types.Allow, types.ImpersonateConditions{
				Roles: req.Bot.Spec.Roles,
			})
```

It is dangerous to use, too: looking at other types using it, the clients would typically calculate the "patch" that needs applying and send it over; only the "patch" is calculated based off their own view of the object state, which may be stale. To be safe, you should do db-style transactions with retries, where client logic only operates on a consistent view of the object.

This kind of call makes more sense for editing UIs where the person performing the edits doesn't care that unrelated parts of the object has been changed, although even then it should be handled with care.

Reply:
@strideynet what are your thoughts on `update_mask` after implementing necessary changes for bots?

Reply:
They can be a tad repetitive to implement - but, the benefits have been fairly clear. i.e the frontend no longer needs to make a Get before it Updates the Bot when looking to simply update the roles.

I think realistically long term we'd be wanting to investigate something that can leverage proto message reflection to automatically apply the update mask.

Reply:
Cool, makes sense in the context of UI, thanks for the context. Does it properly handle the case where two people are editing the set of roles for the bot?

Reply:
Not quite - but I wouldn't take the Bot resource as an example here. It's not persisted to the backend at the moment and is on-the-spot reconciled to and from a Bot User and Bot Role - this isn't the long-term plan but is part of a transition as historically we haven't had a Bot resource (until v15). This somewhat limits my ability to use the current optimistic locking.

Reply:
Update masks are good in situations where not every update is a full update. I find that this happens frequently.

I'm a bit torn about calculating the mask on the fly, as that somewhat defeats the advantage. Fixed masks work better in my opinion (as in, I know this will always and only ever touch fields _x_ and _y_, so these are the masks used).

Another advantage is that you may not need a previous read, like Noah said.

> Does it properly handle the case where two people are editing the set of roles for the bot?

That is not what the masks are for, although they can help "solve" those problems if the concurrent edits are independent and touch disjointed fields. I'd look at optimistic concurrency in this case, which is something that we also miss in Teleport designs.

Reply:
Just to be clear about expectations, I'm not holding the PR for the update_mask, but I would strongly recommend you added it (for whatever that is worth).

Reply:
I gave it some further thought. While I recognise the potential benefits of `update_mask`, these are irrelevant to the current needs of this API. It is a kind of optimisation, which I believe to be premature at this moment. I'm certainly open to revisiting this in the future if needed; the request can be easily extended to accommodate that.

As for the general idea of `update_mask`, it would be quite interesting to see an end-to-end implementation which accounts for the complexity I've mentioned above.